jina/peapods/peas/__init__.py

```diff
@@ -2,102 +2,30 @@
+from contextlib import ExitStack
 from typing import Dict, Optional, Union, List
 
 import zmq
 
-from ..zmq import send_ctrl_message, Zmqlet, ZmqStreamlet
+from ..zmq import Zmqlet, ZmqStreamlet
 from ... import Message
-from ... import __ready_msg__, __stop_msg__, Request
+from ... import Request
 from ...enums import PeaRoleType, SkipOnErrorType
 from ...excepts import RequestLoopEnd, NoExplicitMessage, ExecutorFailToLoad, MemoryOverHighWatermark, DriverError, \
-    PeaFailToStart, \
     ChainedPodException
 from ...executors import BaseExecutor
-from ...helper import is_valid_local_config_source, typename
+from ...helper import is_valid_local_config_source
 from ...logging import JinaLogger
 from ...logging.profile import used_memory, TimeDict
 from ...proto import jina_pb2
 
-__all__ = ['PeaMeta', 'BasePea']
-
-
-class PeaMeta(type):
-    """Meta class of :class:`BasePea` to enable switching between ``thread`` and ``process`` backend. """
-    _dct = {}
-
-    def __new__(cls, name, bases, dct):
-        _cls = super().__new__(cls, name, bases, dct)
-        PeaMeta._dct.update({name: {'cls': cls,
-                                    'name': name,
-                                    'bases': bases,
-                                    'dct': dct}})
-        return _cls
-
-    def __call__(cls, *args, **kwargs) -> 'PeaMeta':
-        # switch to the new backend
-        _cls = {
-            'thread': threading.Thread,
-            'process': multiprocessing.Process,
-        }.get(getattr(args[0], 'runtime', 'thread'))
-
-        # rebuild the class according to mro
-        for c in cls.mro()[-2::-1]:
-            arg_cls = PeaMeta._dct[c.__name__]['cls']
-            arg_name = PeaMeta._dct[c.__name__]['name']
-            arg_dct = PeaMeta._dct[c.__name__]['dct']
-            _cls = super().__new__(arg_cls, arg_name, (_cls,), arg_dct)
-
-        return type.__call__(_cls, *args, **kwargs)
-
-
-def _get_event(obj: 'BasePea') -> Event:
-    if isinstance(obj, threading.Thread):
-        return threading.Event()
-    elif isinstance(obj, multiprocessing.Process):
-        return multiprocessing.Event()
-    else:
-        raise NotImplementedError
-
-
-def _make_or_event(obj: 'BasePea', *events) -> Event:
-    or_event = _get_event(obj)
-
-    def or_set(self):
-        self._set()
-        self.changed()
-
-    def or_clear(self):
-        self._clear()
-        self.changed()
-
-    def orify(e, changed_callback):
-        e._set = e.set
-        e._clear = e.clear
-        e.changed = changed_callback
-        e.set = lambda: or_set(e)
-        e.clear = lambda: or_clear(e)
-
-    def changed():
-        bools = [e.is_set() for e in events]
-        if any(bools):
-            or_event.set()
-        else:
-            or_event.clear()
-
-    for e in events:
-        orify(e, changed)
-    changed()
-    return or_event
+__all__ = ['BasePea']
 
 
-class BasePea(metaclass=PeaMeta):
+class BasePea(ExitStack):
```

Comment:
Is the plan - BasePea `has-a` Runtime rather than `is-a` Runtime?

Reply:
Runtime has a `BasePea`

Reply:
so BasePea does not care about wether it lives in a different process or anything, it just starts the zmq loop and the executor